### ISSUE TITLE
Set as Default for Postgres Databases fails

### ISSUE THREAD

**Issue by yrimal** [dbeaver/dbeaver#7952]
#### System information: 
- Windows 10 Pro Version 2004 (Build 19041.84)
- 6.3.5.202002151353

#### Connection specification:
- PostgreSql 9.6
- PostgreSQL standard driver
- Do you use tunnels or proxies (SSH, SOCKS, etc) : SSH 

#### Describe the problem you're observing:
Switching to another database is clumsy and fails most of the time with error below: 
```
SQL Error [08001]: Connection to 127.0.0.1:15045 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
``` 
However, there is no connection issue and it was working fine on previous version. That was 6.2.4 or something. 

#### Include any warning/errors/backtraces from the logs
```
org.postgresql.util.PSQLException: Connection to 127.0.0.1:16042 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
	at org.postgresql.Driver$ConnectThread.getResult(Driver.java:403)
	at org.postgresql.Driver.connect(Driver.java:261)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.lambda$0(JDBCDataSource.java:157)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:174)
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.openConnection(PostgreDataSource.java:311)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:91)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:80)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.getConnection(JDBCExecutionContext.java:181)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCConnectionImpl.getOriginal(JDBCConnectionImpl.java:60)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCConnectionImpl.cancelBlock(JDBCConnectionImpl.java:693)
	at org.jkiss.dbeaver.model.runtime.BlockCanceler.cancelBlock(BlockCanceler.java:36)
	at org.jkiss.dbeaver.model.runtime.AbstractJob$JobCanceler.run(AbstractJob.java:221)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.net.ConnectException: Connection refused: connect
	at java.base/java.net.PlainSocketImpl.waitForConnect(Native Method)
	at java.base/java.net.PlainSocketImpl.socketConnect(Unknown Source)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)
	at java.base/java.net.AbstractPlainSocketImpl.connect(Unknown Source)
	at java.base/java.net.SocksSocketImpl.connect(Unknown Source)
	at java.base/java.net.Socket.connect(Unknown Source)
	at org.postgresql.core.PGStream.<init>(PGStream.java:69)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:158)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)
	at org.postgresql.Driver.makeConnection(Driver.java:452)
	at org.postgresql.Driver.access$100(Driver.java:61)
	at org.postgresql.Driver$ConnectThread.run(Driver.java:362)
	at java.base/java.lang.Thread.run(Unknown Source)
!SUBENTRY 3 org.jkiss.dbeaver.model 4 0 2020-02-19 11:41:40.873
!MESSAGE Connection refused: connect
!STACK 0
java.net.ConnectException: Connection refused: connect
	at java.base/java.net.PlainSocketImpl.waitForConnect(Native Method)
	at java.base/java.net.PlainSocketImpl.socketConnect(Unknown Source)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)
	at java.base/java.net.AbstractPlainSocketImpl.connect(Unknown Source)
	at java.base/java.net.SocksSocketImpl.connect(Unknown Source)
	at java.base/java.net.Socket.connect(Unknown Source)
	at org.postgresql.core.PGStream.<init>(PGStream.java:69)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:158)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)
	at org.postgresql.Driver.makeConnection(Driver.java:452)
	at org.postgresql.Driver.access$100(Driver.java:61)
	at org.postgresql.Driver$ConnectThread.run(Driver.java:362)
	at java.base/java.lang.Thread.run(Unknown Source)
!SESSION 2020-02-19 11:43:06.566 -----------------------------------------------
eclipse.buildId=unknown
java.version=11.0.5
java.vendor=AdoptOpenJDK
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_US
Command-line arguments:  -os win32 -ws win32 -arch x86_64

!ENTRY org.eclipse.jface 2 0 2020-02-19 11:43:22.529
!MESSAGE Keybinding conflicts occurred.  They may interfere with normal accelerator operation.
!SUBENTRY 1 org.eclipse.jface 2 0 2020-02-19 11:43:22.529
!MESSAGE A conflict occurred for CTRL+N:
Binding(CTRL+N,
	ParameterizedCommand(Command(org.eclipse.ui.newWizard,New,
		Open the New item wizard,
		Category(org.eclipse.ui.category.file,File,null,true),
		org.eclipse.ui.internal.WorkbenchHandlerServiceHandler@3da55998,
		[Parameter(newWizardId,New Wizard,null,true)],,true),null),
	org.eclipse.ui.defaultAcceleratorConfiguration,
	org.eclipse.ui.contexts.window,,,system)
Binding(CTRL+N,
	ParameterizedCommand(Command(org.jkiss.dbeaver.core.sql.editor.open,SQL Editor,
		Open SQL editor (existing or new),
		Category(org.jkiss.dbeaver.core.database,Connection,Database Connection Commands,true),
		org.eclipse.ui.internal.WorkbenchHandlerServiceHandler@40ebb56,
		,,true),null),
	org.jkiss.dbeaver.defaultKeyScheme,
	org.jkiss.dbeaver.ui.perspective,,,user)
Binding(CTRL+N,
	ParameterizedCommand(Command(org.jkiss.dbeaver.core.sql.editor.create,New SQL Editor,
		Open new SQL editor (create new script),
		Category(org.jkiss.dbeaver.core.database,Connection,Database Connection Commands,true),
		org.eclipse.ui.internal.WorkbenchHandlerServiceHandler@5b2fc2c0,
		,,true),null),
	org.jkiss.dbeaver.defaultKeyScheme,
	org.jkiss.dbeaver.ui.perspective,,,user)
```

**Comment by yrimal** [comment]
![image](https://user-images.githubusercontent.com/6102813/74815425-50a15d00-5321-11ea-88e2-ffbeb7f34250.png)

**Comment by uslss** [comment]
thanks for the bug report.

**Comment by serge-rider** [comment]
I still can't reproduce the problem.
But I've found (and fixed) a bug in connection invalidation procedure.

**Comment by yrimal** [comment]
You may need to have multiple databases (not couple) like mine in that server. I can try testing it if the build is available somewhere.

**Comment by uslss** [comment]
@yrimal , EA version with fix is available here https://dbeaver.io/files/ea/

**Comment by yrimal** [comment]
@uslss , The issue still persists on the latest version (7.0.). I was able to replicate it the following way : 

- Open query editor with database_A Set as Default and execute some query. 
- Change the Default database to database_B from Database Navigator and execute some query. 
- Now, from database navigator, change Default database to database_c from database navigator and try to execute some query.  Then you get the error like below: 
```
SQL Error [08001]: Connection to 127.0.0.1:12972 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
```

**Comment by serge-rider** [comment]
The bug was in SSH tunnel invalidation. We must keep original port number.

**Comment by uslss** [comment]
verified

**Comment by uslss** [comment]
Reopened. Still appears if dbeaver was left opened for a long (more than 1 hour) time) when trying to open sql editor
![image](https://user-images.githubusercontent.com/31996417/75552486-74625280-5a47-11ea-89a1-ba6eff436ee6.png)

**Comment by serge-rider** [comment]
Does it happen only once or connection is completely broken?

**Comment by uslss** [comment]
Probably it is completely broken, invalidate doesn't help, only disconnect and then connect to db does

**Comment by serge-rider** [comment]
@uslss please try to reproduce it by putting your machine in sleep mode or by switching to another wifi or by plugging off the network.

**Comment by uslss** [comment]
can't reproduce it anymore in 7.0.1

**Comment by yrimal** [comment]
Will let you know once I test @uslss

**Comment by yrimal** [comment]
This issue appears to have fixed on Version `7.0.1.202003211804` but with some other trade off. When switching Active data object, it's taking longer time compared to version 6.2.4. Some good amount of time is spent on `Obtaining exclusive data sources lock` .

CC: @serge-rider @uslss

**Comment by uslss** [comment]
Thanks for response @yrimal .
I will close this issue as solved.
Please create new issue if it is founded

**Comment by yrimal** [comment]
@uslss ,
I'm concerned with the tag `Can't reproduce`. Since it can be reproduced in the version I originally mentioned, I don't think the label should be `Can't reproduce` .
